In [1]:
import numpy as np
import pandas as pd

### Задачи 1 и 2

#### Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Случайная величина X задаёт число правильных ответов при простом угадывании. Найдите математическое ожидание данной случайной величины. Найдите дисперсию случайной величины X.

#### Подсказка: постройте закон распределения случайной величины X.

In [2]:
def combinations(n: int, k: int) -> int:
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

In [3]:
n = 5 # число вопросов
p = 1 / 4 # вероятность угадать правильный ответ в вопросе из заданных 4

In [4]:
p_0 = combinations(5, 0) * p ** 0 * (1 - p) ** (n - 0) # вероятность не угадать ни одного ответа из 5 вопросов
p_0

0.2373046875

In [5]:
p_1 = combinations(5, 1) * p ** 1 * (1 - p) ** (n - 1) # вероятность угадать 1 ответ из 5 вопросов
p_1

0.3955078125

In [6]:
p_2 = combinations(5, 2) * p ** 2 * (1 - p) ** (n - 2) # вероятность угадать 2 ответа из 5 вопросов
p_2

0.263671875

In [7]:
p_3 = combinations(5, 3) * p ** 3 * (1 - p) ** (n - 3) # вероятность угадать 3 ответа из 5 вопросов
p_3

0.087890625

In [8]:
p_4 = combinations(5, 4) * p ** 4 * (1 - p) ** (n - 4) # вероятность угадать 4 ответа из 5 вопросов
p_4

0.0146484375

In [9]:
p_5 = combinations(5, 5) * p ** 5 * (1 - p) ** (n - 5) # вероятность угадать все 5 ответов во всех 5 вопросах
p_5

0.0009765625

In [10]:
m = n * p # считаем математическое ожидание биномиального распределения
m

1.25

In [11]:
d = n * p * (1 - p) # считаем дисперсию биномиального распределения
d

0.9375

In [12]:
values = [0, 1, 2, 3, 4, 5] # список значений, касаемых количества правильных ответов
probabilities = [p_0, p_1, p_2, p_3, p_4, p_5] # список посчитанных вероятностей

In [13]:
# биномиальный закон распределения

binomial_distribution = pd.DataFrame({'x(i)': values, 'p(i)': probabilities}).set_index('x(i)').T
binomial_distribution

x(i)         0         1         2         3         4         5
p(i)  0.237305  0.395508  0.263672  0.087891  0.014648  0.000977

### Задача 3

#### Пользователь получает в среднем 10 писем со спамом на свой почтовый ящик за сутки. Найти число N, такое, что с вероятностью 0.95 пользователь получит не более N писем со спамом за текущий день.

In [14]:
def poisson_proba(k: int, lambda_: float) -> float:
    return (lambda_ ** k) * (np.exp(-lambda_)) / np.math.factorial(k)

In [15]:
lambda_ = 10 # среднее количество писем со спамом, которые получает пользователь за сутки
p = 0.95 # вероятность, при которой пользователь получает не более N писем со спамом за сутки

In [16]:
# задано условие, что пользователь получит не более N писем со спамом, значит от 0 до N включительно
# соответственно, вероятности нужно просуммировать

N = range(0, 21) # предположим, что писем не более 20

for n in N:
    if round(sum(poisson_proba(k=i, lambda_=lambda_) for i in range(n)), 2) == p: # ищем такое n, при котором вероятность будет равна p (0.95)
        print(n)
        print()
        print(f'При вероятности {p} пользователь получает не более {n - 1} писем со спамом за сутки')

16

При вероятности 0.95 пользователь получает не более 15 писем со спамом за сутки


#### Получаем 16, а так как range не включает последнее число, то ответом будет 15.

### Задача 4

#### Производятся выстрелы по мишени. Вероятность попадания в мишень при одном выстреле равна 0.01. Сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 6 раз будет совершено попадание?

#### Подсказка: 1) "Вероятность попасть k раз при n выстрелах" - на какое распределение это похоже? 2) А если нам нужна вероятность P(X >= k), а не P(X = k)? 3) Здесь предстоит немножко покодить.

In [17]:
def bernoulli(n: int, k: int, p: float) -> float:
    return combinations(n, k) * p ** k * (1 - p) ** (n - k)

In [18]:
p = 0.01
p_gte_6 = 0.9
p_lt_6 = round(1 - p_gte_6, 1)
p_lt_6

0.1

In [19]:
# вычислим количество выстрелов, которое нужно сделать, чтобы попасть хотя бы 6 раз (6 и более) с вероятностью 0.9

for n in range(5, 1500): # предположим, что выстрелов не более 1500
    if round(sum(bernoulli(n=n, k=i, p=p) for i in range(6)), 1) == p_lt_6:
        if round(1 - sum(bernoulli(n=n, k=i, p=p) for i in range(6)), 1) == p_gte_6:
            print(n)

848
849
850
851
852
853
854
855
856
857
858
859
860
861
862
863
864
865
866
867
868
869
870
871
872
873
874
875
876
877
878
879
880
881
882
883
884
885
886
887
888
889
890
891
892
893
894
895
896
897
898
899
900
901
902
903
904
905
906
907
908
909
910
911
912
913
914
915
916
917
918
919
920
921
922
923
924
925
926
927
928
929
930
931
932
933
934
935
936
937
938
939
940
941
942
943
944
945
946
947
948
949
950
951
952
953
954
955
956
957
958
959
960
961
962
963
964
965
966
967
968
969
970
971
972
973
974
975
976
977
978
979
980
981
982
983
984
985
986
987
988
989
990
991
992
993
994
995
996
997
998
999
1000
1001
1002
1003
1004
1005
1006
1007
1008
1009
1010
1011
1012
1013
1014
1015
1016
1017
1018
1019
1020
1021
1022
1023
1024
1025
1026
1027
1028
1029
1030
1031
1032
1033
1034
1035
1036
1037
1038
1039
1040
1041
1042
1043
1044
1045
1046
1047
1048


#### Проверим крайние значения (848 и 1048):

In [20]:
print(round(sum(bernoulli(n=848, k=i, p=p) for i in range(6)), 1) == p_lt_6) # 0.1
print(round(1 - sum(bernoulli(n=848, k=i, p=p) for i in range(6)), 1) == p_gte_6) # 0.9

True
True


In [21]:
print(round(sum(bernoulli(n=1048, k=i, p=p) for i in range(6)), 1) == p_lt_6) # 0.1
print(round(1 - sum(bernoulli(n=1048, k=i, p=p) for i in range(6)), 1) == p_gte_6) # 0.9

True
True


#### Убедимся, что при значениях 847 и 1049 ответ будет неверным:

In [22]:
print(round(sum(bernoulli(n=847, k=i, p=p) for i in range(6)), 1) == p_lt_6) # !0.1
print(round(1 - sum(bernoulli(n=847, k=i, p=p) for i in range(6)), 1) == p_gte_6) # !0.9

False
False


In [23]:
print(round(sum(bernoulli(n=1049, k=i, p=p) for i in range(6)), 1) == p_lt_6) # !0.1
print(round(1 - sum(bernoulli(n=1049, k=i, p=p) for i in range(6)), 1) == p_gte_6) # !0.9

False
False


In [24]:
# если не округлять значения, то самое близкое количество выстрелов к вероятностям 0.1 и 0.9 - это 925

print(sum(bernoulli(n=925, k=i, p=p) for i in range(6)))
print(1 - sum(bernoulli(n=925, k=i, p=p) for i in range(6)))

0.10017581020633598
0.8998241897936641


#### В ответе получаем, что для попадания хотя бы 6 раз (6 и более) с вероятностью 0.9 нужно произвести от 848 до 1048 выстрелов.

#### Если посчитать среднее, то в ответ можно записать 948 (выстрелов).

#### 925 выстрелов - самое близкое количество выстрелов к вероятностям 0.1 и 0.9, если не округлять значения.